File: android/app/src/main/java/au/com/refrigerationservices/weather/data/Recipients.kt
package au.com.refrigerationservices.weather.data

/**
 * The recipient rule, ported from server/src/recipients.js.
 *
 * This copy exists so the UI can reject a bad address instantly without a round
 * trip. It is NOT the enforcement point — the server re-checks every address
 * before anything is sent. server/test/shared-rule.test.js fails if this port
 * drifts from the server's domain, wording or anchoring.
 */
object Recipients {

    const val ALLOWED_DOMAIN = "refrigerationservices.com.au"

    /** The exact message the business asked for when an address fails the rule. */
    const val REJECTION_MESSAGE = "This email is not one of your family member"

    /**
     * Anchored at both ends so a lookalike such as
     * `a@refrigerationservices.com.au.evil.com` cannot pass. The local part
     * excludes whitespace, `@`, commas and semicolons so a recipient list can
     * never be smuggled through as a single address.
     */
    private val ADDRESS = Regex(
        "^[^\\s@,;<>\"]+@refrigerationservices\\.com\\.au$",
        RegexOption.IGNORE_CASE,
    )

    fun isAllowed(email: String?): Boolean =
        email != null && ADDRESS.matches(email.trim())

    /** Validate and normalise, or explain the refusal. */
    fun check(email: String?): Result {
        val trimmed = email?.trim().orEmpty()
        return when {
            trimmed.isEmpty() -> Result.Invalid("Enter an email address first.")
            !isAllowed(trimmed) -> Result.Invalid(REJECTION_MESSAGE)
            else -> Result.Valid(trimmed.lowercase())
        }
    }

    sealed interface Result {
        data class Valid(val email: String) : Result
        data class Invalid(val error: String) : Result
    }
}

File: server/src/recipients.js
/**
 * The recipient rule: reports may only ever go to @refrigerationservices.com.au.
 *
 * This lives in its own module because it is enforced in three places — the web
 * UI, the Android UI, and the chat agent's send tool. The UI checks exist to give
 * fast feedback; the server-side check in the send path is the one that actually
 * enforces the rule, since chat input is untrusted and a model can be talked into
 * trying to send somewhere else.
 */

export const ALLOWED_DOMAIN = 'refrigerationservices.com.au';

/** The exact message the user asked for when an address fails the rule. */
export const REJECTION_MESSAGE = 'This email is not one of your family member';

/**
 * Anchored on both ends so nothing like `a@refrigerationservices.com.au.evil.com`
 * or `a@evil.com?x=@refrigerationservices.com.au` can slip through. The local part
 * excludes whitespace, `@` and commas so a comma-separated list can never be
 * smuggled in as a single "address".
 */
const ADDRESS_RE = /^[^\s@,;<>"]+@refrigerationservices\.com\.au$/i;

/** @returns {boolean} whether this address may receive a report. */
export function isAllowedRecipient(email) {
  if (typeof email !== 'string') return false;
  return ADDRESS_RE.test(email.trim());
}

/**
 * Validate and normalise, or explain the refusal.
 * @returns {{ ok: true, email: string } | { ok: false, error: string }}
 */
export function checkRecipient(email) {
  const trimmed = typeof email === 'string' ? email.trim() : '';
  if (!trimmed) return { ok: false, error: 'Enter an email address first.' };
  if (!isAllowedRecipient(trimmed)) return { ok: false, error: REJECTION_MESSAGE };
  return { ok: true, email: trimmed.toLowerCase() };
}
